18 implement chartjs

simo2898 opened the pull request.

simo2898 committed:
Added chart for expenses

simo2898 committed:
Added buttons to move between expenses month charts

simo2898 committed:
Added api that returns all the expenses by year
Added chart that shows all the expenses per year

mattiapitossi commented on src/main/java/com/expensetracker/controller/ExpenseController.java:
```diff
@@ -32,6 +33,12 @@ ResponseEntity<List<Expense>> getExpensesBetween(@RequestParam(name = "month") i
 
+
+    @GetMapping("/expenses/{year}")
+    ResponseEntity<List<YearExpensesDTO>> getExpensesBetween(@PathVariable(name = "year") int year) {
```
I'd add a validation to the year, what if we receive something like 2?

mattiapitossi commented on src/main/java/com/expensetracker/model/YearExpensesDTO.java:
```diff
@@ -0,0 +1,17 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Data
+public class YearExpensesDTO {
```
Since we're using Java 17, I'd use a record here to encapsulate data classes

simo2898 replied:
Changed YearExpensesDTO from a class to a record.

mattiapitossi commented on src/main/java/com/expensetracker/service/impl/ExpenseServiceImpl.java:
```diff
@@ -68,6 +69,16 @@ public List<Expense> getAllExpensesPeriodByDate(int month, int year) {
+        LocalDate startDate = localDate.with(TemporalAdjusters.firstDayOfYear());
+        LocalDate endDate = localDate.with(TemporalAdjusters.lastDayOfYear());
+
+        System.out.println(expenseRepository.getExpensesByYear(startDate, endDate));
```
Wouldn't be better to use a logger?

simo2898 replied:
Added issue to add a logger and removed all the print methods in the backend code

simo2898 committed:
Merge remote-tracking branch 'origin/20-migrate-from-react-to-vue' into 18-implement-chartjs

# Conflicts:
#	src/main/java/com/expensetracker/service/ExpenseService.java
#	src/main/java/com/expensetracker/service/impl/ExpenseServiceImpl.java

simo2898 committed:
Changed YearExpensesDTO from a class to a record
Removed print methods
Added Chart files

simo2898 committed:
Added constraints to year and month

simo2898 committed:
Improve UI interface
Modified subcategory option to contain only subcategory linked to the category and created a method that return a subcategory list from a category's name.
Added methods for populating charts data.
Modified old db diagram

mattiapitossi commented on src/main/java/com/expensetracker/controller/ExpenseController.java:
```diff
@@ -27,11 +31,23 @@ ResponseEntity<List<Expense>> getExpenses() {
-    ResponseEntity<List<Expense>> getExpensesBetween(@RequestParam(name = "month") int month,
-                                                     @RequestParam(name = "year") int year) {
+    ResponseEntity<List<Expense>> getExpensesBetween(@RequestParam(name = "month") @Range(min = 1, max = 12) int month,
+                                                     @RequestParam(name = "year") @Range(min = 2020, max = 2099) int year) {
```
What if I want to track an expense previous to 2020?

mattiapitossi commented on src/main/java/com/expensetracker/controller/SubcategoryController.java:
```diff
@@ -24,6 +24,11 @@ ResponseEntity<List<Subcategory>> readSubcategories() {
         return ResponseEntity.ok().body(subcategoryService.getAllSubcategories());
     }
 
+    @GetMapping("/subcategory/{id}")
```
```suggestion
    @GetMapping("/subcategories/{id}")
```

mattiapitossi commented on src/main/java/com/expensetracker/model/dto/YearExpensesDTO.java:
```diff
@@ -0,0 +1,4 @@
+package com.expensetracker.model.dto;
+
+public record YearExpensesDTO(Object date, Object value, Object transaction) {
```
Object? Wouldn't be better to use LocalDate?

simo2898 replied:
Spring has problems converting the month function to a LocalDate or a String.

mattiapitossi commented on src/main/java/com/expensetracker/repository/SubcategoryRepository.java:
```diff
@@ -5,9 +5,12 @@
 
     Subcategory findByNameAndCategory(String name, Category category);
 
+    List<Subcategory> findByCategory_Name(String categoryName);
```
Better to follow the Java methods name convention?

mattiapitossi commented on src/main/java/com/expensetracker/service/ExpenseService.java:
```diff
@@ -19,6 +22,10 @@ public interface ExpenseService {
 
+    List<YearExpensesDTO> getAllExpensesByYear(int year);
+
+    List<CategoryExpensesDTO> getCategoriesExpensesByMonth(int month, int year);
```
```suggestion
    List<CategoryExpensesDTO> getCategoriesExpensesBy(int month, int year);
```

mattiapitossi commented on src/main/java/com/expensetracker/service/ExpenseService.java:
```diff
@@ -19,6 +22,10 @@ public interface ExpenseService {
 
     List<Expense> getAllExpensesPeriodByDate(int month, int year);
 
+    List<YearExpensesDTO> getAllExpensesByYear(int year);
```
```suggestion
    List<YearExpensesDTO> getAllExpensesBy(int year);
```

mattiapitossi commented on src/main/java/com/expensetracker/service/SubcategoryService.java:
```diff
@@ -21,6 +21,11 @@ public List<Subcategory> getAllSubcategories() {
 
+
+    public List<Subcategory> getAllSubcategoriesFromCategoryName(String categoryName) {
+        return subcategoryRepository.findByCategory_Name(categoryName);
```
Same as before, why the underscore?

mattiapitossi commented on src/main/java/com/expensetracker/service/impl/ExpenseServiceImpl.java:
```diff
@@ -68,6 +70,24 @@ public List<Expense> getAllExpensesPeriodByDate(int month, int year) {
     }
 
+    @Override
+    public List<YearExpensesDTO> getAllExpensesByYear(int year) {
```
```suggestion
    public List<YearExpensesDTO> getAllExpensesBy(int year) {
```

mattiapitossi commented on src/main/java/com/expensetracker/service/impl/ExpenseServiceImpl.java:
```diff
@@ -68,6 +70,24 @@ public List<Expense> getAllExpensesPeriodByDate(int month, int year) {
+        LocalDate startDate = localDate.with(TemporalAdjusters.firstDayOfYear());
+        LocalDate endDate = localDate.with(TemporalAdjusters.lastDayOfYear());
+
+        return expenseRepository.getExpensesByYear(startDate, endDate);
```
```suggestion
        return expenseRepository.getExpensesBy(startDate, endDate);
```

mattiapitossi commented on src/main/java/com/expensetracker/service/impl/ExpenseServiceImpl.java:
```diff
@@ -68,6 +70,24 @@ public List<Expense> getAllExpensesPeriodByDate(int month, int year) {
+    }
+
+    @Override
+    public List<CategoryExpensesDTO> getCategoriesExpensesByMonth(int month, int year) {
```
```suggestion
    public List<CategoryExpensesDTO> getCategoriesExpensesBy(int month, int year) {
```

mattiapitossi commented on src/main/java/com/expensetracker/service/impl/ExpenseServiceImpl.java:
```diff
@@ -68,6 +70,24 @@ public List<Expense> getAllExpensesPeriodByDate(int month, int year) {
+        LocalDate startDate = localDate.with(TemporalAdjusters.firstDayOfMonth());
+        LocalDate endDate = localDate.with(TemporalAdjusters.lastDayOfMonth());
+
+        return expenseRepository.getCategoryExpensesByMonth(startDate, endDate);
```
```suggestion
        return expenseRepository.getCategoryExpensesBy(startDate, endDate);
```

simo2898 committed:
Code Review Requested Changes
Renamed "By" Methods
Changed mapping subcategory to subcategories
Remove underscore from repository's methods
